internal/crypto/cert.go

```diff
@@ -37,32 +48,104 @@ func CNFromDeviceFingerprint(fingerprint string) (string, error) {
 	return DeviceCommonNamePrefix + fingerprint, nil
 }
 
+func CertFile(name string) string {
+	return filepath.Join(config.CertificateDir(), name+".crt")
+}
+
+func KeyFile(name string) string {
+	return filepath.Join(config.CertificateDir(), name+".key")
+}
+
 type TLSCertificateConfig oscrypto.TLSCertificateConfig
 
-type CA struct {
-	Config *TLSCertificateConfig
+type CA interface {
+	GetId() string
+	GetConfig() *TLSCertificateConfig
+	EnsureServerCertificate(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, bool, error)
+	MakeAndWriteServerCert(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, error)
+	MakeServerCert(hostnames []string, expiryDays int, fns ...CertificateExtensionFunc) (*TLSCertificateConfig, error)
+	signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error)
+	EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error)
+	MakeClientCertificate(certFile, keyFile string, subject string, expiryDays int) (*TLSCertificateConfig, error)
+	IssueRequestedClientCertificate(csr *x509.CertificateRequest, expirySeconds int) ([]byte, error)
+}
+
+
+
+var caMutex = &sync.Mutex{}
+var caList map[string]CA = make(map[string]CA)
+
+func EnsureCA(id string, cfg *config.Config) (CA, bool, error) {
```

Comment:
I think we can reduce the scope of this signature to just the CryptographyConfig.

```suggestion
func EnsureCA(id string, cfg *config.CryptographyConfig) (CA, bool, error) {
```